mongo.go

```diff
@@ -0,0 +1,472 @@
+	}
+	return true, nil
+}
+
+func (m *MongoDS) getSize(ctx context.Context, key datastore.Key) (int, error) {
+	v, err := m.get(ctx, key)
+	if err == datastore.ErrNotFound {
+		return -1, err
+	}
+	if err != nil {
+		return 0, fmt.Errorf("getting value: %s", err)
+	}
+	return len(v), nil
+}
+
+func (m *MongoDS) query(ctx context.Context, q dsextensions.QueryExt) (query.Results, error) {
+	opts := options.Find()
+
+	// Handle ordering
+	asc := true
+	if len(q.Orders) > 0 {
+		switch q.Orders[0].(type) {
+		case dsq.OrderByKey, *dsq.OrderByKey:
+			// We order by key by default.
+		case dsq.OrderByKeyDescending, *dsq.OrderByKeyDescending:
+			// Reverse order by key
+			asc = false
+		default:
+			// Ok, we have a weird order we can't handle. Let's
+			// perform the _base_ query (prefix, filter, etc.), then
+			// handle sort/offset/limit later.
+
+			// Skip the stuff we can't apply.
+			baseQuery := q
+			baseQuery.Limit = 0
+			baseQuery.Offset = 0
+			baseQuery.Orders = nil
+
+			// perform the base query.
+			res, err := m.query(ctx, baseQuery)
+			if err != nil {
+				return nil, err
+			}
+
+			// fix the query
+			res = dsq.ResultsReplaceQuery(res, q.Query)
+
+			// Remove the parts we've already applied.
+			naiveQuery := q.Query
+			naiveQuery.Prefix = ""
+			naiveQuery.Filters = nil
+
+			// Apply the rest of the query
+			return dsq.NaiveQueryApply(naiveQuery, res), nil
+		}
+	}
+	opts.SetSort(bson.M{"_id": 1})
+	if !asc {
+		opts.SetSort(bson.M{"_id": -1})
+	}
+
+	prefix := datastore.NewKey(q.Prefix).String()
+	// Important to consider the '/' suffix to respect Prefix semantics
+	// of returning strictly child keys.
+	var filters bson.A
+	if prefix != "/" {
+		rgx := fmt.Sprintf("^%s/.*", prefix)
+		filters = append(filters, bson.M{"_id": bson.M{"$regex": primitive.Regex{Pattern: rgx}}})
+	}
+	seekPrefix := datastore.NewKey(q.SeekPrefix).String()
+	if seekPrefix != "/" {
+		op := "$gte"
+		if !asc {
+			op = "$lte"
+		}
+		filters = append(filters, bson.M{"_id": bson.M{op: seekPrefix}})
+	}
+	fil := bson.M{}
+	if len(filters) > 0 {
+		fil = bson.M{"$and": filters}
+	}
+
+	// If we have no filters, then we can leverage Skip.
+	// If that isn't the case, we should fetch all of them
+	// and apply skipping later.
+	if len(q.Filters) == 0 {
+		opts.SetSkip(int64(q.Offset))
+	}
+
+	it, err := m.col.Find(ctx, fil, opts)
+	if err != nil {
+		return nil, fmt.Errorf("finding key-values: %s", err)
+	}
+
+	qrb := dsq.NewResultBuilder(q.Query)
+	qrb.Process.Go(func(worker goprocess.Process) {
+		m.lock.RLock()
+		closedEarly := false
+		defer func() {
+			m.lock.RUnlock()
+			if closedEarly {
+				select {
+				case qrb.Output <- dsq.Result{
+					Error: ErrClosed,
+				}:
+				case <-qrb.Process.Closing():
+				}
+			}
+
+		}()
+		if m.closed {
+			closedEarly = true
+			return
+		}
+
+		ctx, cls := context.WithTimeout(context.Background(), m.opTimeout)
+		defer cls()
+		defer it.Close(ctx)
+
+		if len(q.Filters) > 0 {
```

Comment:
Do the worst case scenario of skipping manually if we need to. If filters were empty, we already solved the skip with `SetSkip` in opts.